INTRODUCTION AND DATA EXPLORATION

To perform data cleaning and exploration, we will apply quiet a number of techniques using the various libraries as contained in python. The code below will help us read out data into the environment.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Loading the climate data file
from csv import reader

# Load a CSV file
def load_csv(filename):
    file = open(filename, "r")
    lines = reader(file)
    dataset = list(lines)
    return dataset

# Load dataset
filename = 'fueldatapy.csv'
dataset = load_csv(filename)
print('Loaded data file {0} with {1} rows and {2} columns'.format(filename, len(dataset),
len(dataset[0])))

Loaded data file fueldatapy.csv with 29524 rows and 11 columns


From the above we observe that out data contain 29524 rows with 11 columns.  To explore the features available in our dataset we will check the following parameters:
i. Check the colunm names of the dataset for posible need to recode.
ii. Top 5 rows of the dataset to have a glimpse of the data we have at hand.
iii. Detrmine the data type of the variables stored in each column.
iv. Priliminary explore the summary statistics to cover Mean, median, and other statistical attributes.
v. Determine the unique elements in each row.

These and other steps will be taken to pre-process and understand the nature of data we have. The following code below enables us to check the colunm names of our data set.

In [2]:
# importing the csv library
import csv

# list to store the names of columns
list_of_column_names = []

# loop to iterate thorugh the rows of csv
for row in dataset:
    list_of_column_names.append(row)# adding the first row
    break

# printing the result
print("List of column names : ",
      list_of_column_names[0])


List of column names :  ['record_id', 'utility_id_ferc1', 'report_year', 'plant_name_ferc1', 'fuel_type_code_pudl', 'fuel_unit', 'fuel_qty_burned', 'fuel_mmbtu_per_unit', 'fuel_cost_per_unit_burned', 'fuel_cost_per_unit_delivered', 'fuel_cost_per_mmbtu']


From the above, it is evident that there will be need to rename and recode the colunms. To proceed with checking the top rows of the data set we will use the following code:

In [3]:
# making data frame
fueldata = pd.read_csv("fueldatapy.csv")
  
fueldatatop = fueldata.head() # calling head() method and storing in new variable
  
# display the output
print(fueldatatop)


                record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton          5377489   
1  rockport total plant                coal       ton         10486945   
2                gorgas                coal       ton          2978683   
3                 barry                coal       ton          3739484   
4             chickasaw                 gas       mcf            40533   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130             

This have given us a glimpse of the nature of data we are having. We will also explore to determine the uniqueness of the dataset. The following code will do just that for us.

In [4]:
# summarize the number of unique values in each column
print(fueldata.nunique())

record_id                       29523
utility_id_ferc1                  185
report_year                        25
plant_name_ferc1                 2315
fuel_type_code_pudl                 6
fuel_unit                           9
fuel_qty_burned                 26432
fuel_mmbtu_per_unit             11213
fuel_cost_per_unit_burned       19416
fuel_cost_per_unit_delivered    16675
fuel_cost_per_mmbtu             12590
dtype: int64


DATA CLEANING, FEATURE EXTRACTION AND DATA TRANSFORMATION

We will perform basic data cleaning by removing rows with low varainces, droping colunms that are not useful for computation e.t.c. Similarly, recall when we did find unique values of our data entry, we observed that 'report year' have some string entry, we will also try to clean that up using Numpy and str method.

In [5]:
fueldata["report_year"].unique()

array([1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004,
       2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,
       2016, 2017, 2018], dtype=int64)

In a previour exploration of the downloaded data, we realised that some entries of the data are not properly defined and coded. We will attempt to perform data clean up as we can.  It is indeed evident that we will have to clean up this feature as entry such as '20-Mar', '20-Apr', '20-May', '20-Jun', '20-Jul', '20-Aug', '20-Sep', '20-Oct', '20-Nov', '20-Dec' are typo error and possibly belong to the year 2000. This is done using the following R code:

setwd("~/stageA")
data <- read.csv("fueldata.csv", header = TRUE, stringsAsFactors = FALSE)
View(data)
### CONVERT STRING DATA ENTRY TO THE APPROPRIATE FORMAT
data[data == "20-Jan"] <- "2001"
data[data == "20-Feb"] <- "2002"
data[data == "20-Mar"] <- "2003"
data[data == "20-Apr"] <- "2004"
data[data == "20-May"] <- "2005"
data[data == "20-Jun"] <- "2006"
data[data == "20-Jul"] <- "2007"
data[data == "20-Aug"] <- "2008"
data[data == "20-Sep"] <- "2009"
data[data == "20-Oct"] <- "2010"
data[data == "20-Nov"] <- "2011"
data[data == "20-Dec"] <- "2012"

write.csv(data,'fueldatapy.csv')


From the above result, we can see that our data frame have very unique data entry. It is of great interest to explore the uniqueness of individual features.

Sequal to our previous discovery, there is need to understand the nature and class of the dataframe. This is acheived below. We have about 29532 entries of eleven (11) columns. The feature "record_id" which is an object class will have to be remove. It is interesting to know that we do not have any empty entry or single valued variable. In summary, we have 64 float data type, two interger data types and five object data types.

In [6]:
#determine the data types
print(fueldata.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
record_id                       29523 non-null object
utility_id_ferc1                29523 non-null int64
report_year                     29523 non-null int64
plant_name_ferc1                29523 non-null object
fuel_type_code_pudl             29523 non-null object
fuel_unit                       29343 non-null object
fuel_qty_burned                 29523 non-null int64
fuel_mmbtu_per_unit             29523 non-null float64
fuel_cost_per_unit_burned       29523 non-null float64
fuel_cost_per_unit_delivered    29523 non-null float64
fuel_cost_per_mmbtu             29523 non-null float64
dtypes: float64(4), int64(3), object(4)
memory usage: 2.5+ MB
None


With the above information gathered so far about our data, it will be of great interest to check the descriptives.  It is observed that NAN was returned for objects that are string. We will probe further by making some data cleaning to have a better overiview.

In [7]:
# importing regex module
import re

# Percentiles to calculate
pert =[.10, .25, .50, .75, .90, .99]
  
# Data types to be used
include =['object', 'float', 'int']
  
# Describing the data
fueldesc = fueldata.describe(percentiles = pert, include = include, exclude = None)
  
# display
print(fueldesc)

                       record_id plant_name_ferc1 fuel_type_code_pudl  \
count                      29523            29523               29523   
unique                     29523             2315                   6   
top     f1_fuel_2014_12_147_0_10        big stone                 gas   
freq                           1              156               11486   
mean                         NaN              NaN                 NaN   
std                          NaN              NaN                 NaN   
min                          NaN              NaN                 NaN   
10%                          NaN              NaN                 NaN   
25%                          NaN              NaN                 NaN   
50%                          NaN              NaN                 NaN   
75%                          NaN              NaN                 NaN   
90%                          NaN              NaN                 NaN   
99%                          NaN              NaN  

In [8]:
#droping columns not needed for computation

newfueldata = fueldata.drop('record_id', axis=1)

print(newfueldata.shape)

(29523, 10)


We will as well perform cleaning by removing quotation marks and other string typographicals that might not be need for computation. 

In [10]:
fueldes = fueldata.describe(percentiles = pert, include = include, exclude = None)
print(fueldes)

                       record_id plant_name_ferc1 fuel_type_code_pudl  \
count                      29523            29523               29523   
unique                     29523             2315                   6   
top     f1_fuel_2014_12_147_0_10        big stone                 gas   
freq                           1              156               11486   
mean                         NaN              NaN                 NaN   
std                          NaN              NaN                 NaN   
min                          NaN              NaN                 NaN   
10%                          NaN              NaN                 NaN   
25%                          NaN              NaN                 NaN   
50%                          NaN              NaN                 NaN   
75%                          NaN              NaN                 NaN   
90%                          NaN              NaN                 NaN   
99%                          NaN              NaN  

In [15]:
fueldata.skew()

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64

In [16]:
fueldata.kurtosis()

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64

In [17]:
fueldata.isnull()

record_id  utility_id_ferc1  report_year  plant_name_ferc1  \
0          False             False        False             False   
1          False             False        False             False   
2          False             False        False             False   
3          False             False        False             False   
4          False             False        False             False   
5          False             False        False             False   
6          False             False        False             False   
7          False             False        False             False   
8          False             False        False             False   
9          False             False        False             False   
10         False             False        False             False   
11         False             False        False             False   
12         False             False        False             False   
13         False             False        False             False   
14         False             False        False             False   
15         False             False        False             False   
16         False             False        False             False   
17         False             False        False             False   
18         False             False        False             False   
19         False             False        False             False   
20         False             False        False             False   
21         False             False        False             False   
22         False             False        False             False   
23         False             False        False             False   
24         False             False        False             False   
25         False             False        False             False   
26         False             False        False             False   
27         False             False        False             False   
28         False             False        False             False   
29         False             False        False             False   
...          ...               ...          ...               ...   
29493      False             False        False             False   
29494      False             False        False             False   
29495      False             False        False             False   
29496      False             False        False             False   
29497      False             False        False             False   
29498      False             False        False             False   
29499      False             False        False             False   
29500      False             False        False             False   
29501      False             False        False             False   
29502      False             False        False             False   
29503      False             False        False             False   
29504      False             False        False             False   
29505      False             False        False             False   
29506      False             False        False             False   
29507      False             False        False             False   
29508      False             False        False             False   
29509      False             False        False             False   
29510      False             False        False             False   
29511      False             False        False             False   
29512      False             False        False             False   
29513      False             False        False             False   
29514      False             False        False             False   
29515      False             False        False             False   
29516      False             False        False             False   
29517      False             False        False             False   
29518      False             False        False             False   
29519

In [18]:
fueldata.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [19]:
fueldata.isnull()

# show the boolean dataframe            
print(" \nshow the boolean Dataframe : \n\n", fueldata.isnull())
  
# Count total NaN at each column in a DataFrame
print(" \nCount total NaN at each column in a DataFrame : \n\n",
      fueldata.isnull().sum())

 
show the boolean Dataframe : 

        record_id  utility_id_ferc1  report_year  plant_name_ferc1  \
0          False             False        False             False   
1          False             False        False             False   
2          False             False        False             False   
3          False             False        False             False   
4          False             False        False             False   
5          False             False        False             False   
6          False             False        False             False   
7          False             False        False             False   
8          False             False        False             False   
9          False             False        False             False   
10         False             False        False             False   
11         False             False        False             False   
12         False             False        False             False   


In [ ]:
fueldata[fuel_unit].isnull/fueldata[fuel_unit].isnull*100

In [21]:
(180/29343)*100

0.6134342091810653

In [ ]:
fueldata[fuel_unit].mean()